File: analyze_semantic_significance.py
#!/usr/bin/env python3
"""
Statistical Significance Analysis of Monster Prime Vectorization
Proves that semantic structure exists in the 71-layer embeddings
"""

import polars as pl
import numpy as np
from scipy import stats
from pathlib import Path
import json

def load_shard(layer: int) -> pl.DataFrame:
    """Load a single layer shard"""
    path = f"vectors_layer_{layer:02}.parquet"
    if Path(path).exists():
        return pl.read_parquet(path)
    return None

def compute_intra_shard_similarity(layer: int) -> dict:
    """
    Compute similarity within a shard (same Monster prime)
    High similarity = semantic clustering by prime
    """
    df = load_shard(layer)
    if df is None or len(df) < 2:
        return None
    
    # Group by shard (Monster prime assignment)
    shard_groups = df.group_by("shard").agg(pl.count())
    
    results = {
        "layer": layer,
        "total_vectors": len(df),
        "num_shards": len(shard_groups),
        "shard_sizes": shard_groups["count"].to_list(),
        "entropy": stats.entropy(shard_groups["count"].to_list()),
    }
    
    return results

def compute_cross_layer_coherence() -> dict:
    """
    Measure if same file/column maintains consistent shard across layers
    High coherence = semantic stability across Monster group structure
    """
    coherence_scores = []
    
    # Sample layers
    layers = [0, 17, 35, 53, 70]  # Spread across 71 layers
    dfs = []
    
    for layer in layers:
        df = load_shard(layer)
        if df is not None:
            dfs.append(df.select(["file", "column", "shard", "layer"]))
    
    if len(dfs) < 2:
        return {"error": "Not enough layers"}
    
    # Join on file+column to track same semantic entity
    base = dfs[0]
    for df in dfs[1:]:
        base = base.join(df, on=["file", "column"], suffix=f"_l{df['layer'][0]}")
    
    # Compute shard consistency
    shard_cols = [c for c in base.columns if c.startswith("shard")]
    if len(shard_cols) > 1:
        # Standard deviation of shard assignments per entity
        shard_std = base.select(shard_cols).to_numpy().std(axis=1).mean()
        coherence = 1.0 / (1.0 + shard_std)  # Lower std = higher coherence
    else:
        coherence = 0.0
    
    return {
        "layers_analyzed": layers,
        "entities_tracked": len(base),
        "shard_consistency_score": float(coherence),
        "interpretation": "High score = same semantic entity stays in same Monster prime shard"
    }

def test_null_hypothesis() -> dict:
    """
    H0: Shard assignments are random (uniform distribution)
    H1: Shard assignments follow semantic structure
    
    Use chi-square test for goodness of fit
    """
    # Aggregate all shard assignments across layers
    all_shards = []
    
    for layer in range(71):
        df = load_shard(layer)
        if df is not None:
            all_shards.extend(df["shard"].to_list())
    
    if len(all_shards) < 100:
        return {"error": "Insufficient data"}
    
    # Observed distribution
    observed = np.bincount(all_shards, minlength=15)
    
    # Expected uniform distribution
    expected = np.full(15, len(all_shards) / 15)
    
    # Chi-square test
    chi2, p_value = stats.chisquare(observed, expected)
    
    return {
        "null_hypothesis": "Shard assignments are uniformly random",
        "alternative_hypothesis": "Shard assignments follow semantic structure",
        "chi_square_statistic": float(chi2),
        "p_value": float(p_value),
        "reject_null": p_value < 0.05,
        "significance_level": 0.05,
        "interpretation": "p < 0.05 means semantic structure exists (reject randomness)",
        "observed_distribution": observed.tolist(),
        "expected_uniform": expected.tolist(),
    }

def analyze_column_type_clustering() -> dict:
    """
    Test if similar column types cluster in same shards
    E.g., numeric columns vs string columns
    """
    # Sample from layer 0
    df = load_shard(0)
    if df is None:
        return {"error": "No data"}
    
    # Heuristic: column names with numbers likely numeric
    df = df.with_columns([
        pl.col("column").str.contains(r"\d").alias("likely_numeric")
    ])
    
    # Compare shard distributions
    numeric_shards = df.filter(pl.col("likely_numeric"))["shard"].to_list()
    text_shards = df.filter(~pl.col("likely_numeric"))["shard"].to_list()
    
    if len(numeric_shards) < 10 or len(text_shards) < 10:
        return {"error": "Insufficient samples"}
    
    # KS test: are distributions different?
    ks_stat, p_value = stats.ks_2samp(numeric_shards, text_shards)
    
    return {
        "test": "Kolmogorov-Smirnov two-sample test",
        "hypothesis": "Numeric and text columns have different shard distributions",
        "ks_statistic": float(ks_stat),
        "p_value": float(p_value),
        "significant_difference": p_value < 0.05,
        "interpretation": "p < 0.05 means column types cluster differently (semantic meaning)",
        "numeric_sample_size": len(numeric_shards),
        "text_sample_size": len(text_shards),
    }

def main():
    print("🔬 STATISTICAL SIGNIFICANCE ANALYSIS")
    print("=" * 70)
    print()
    
    results = {
        "analysis_date": "2026-01-29",
        "method": "Monster Prime Vectorization (71 layers, 15 shards)",
        "tests": {}
    }
    
    # Test 1: Null hypothesis (randomness)
    print("📊 Test 1: Testing for non-random structure...")
    test1 = test_null_hypothesis()
    results["tests"]["randomness_test"] = test1
    
    if "p_value" in test1:
        if test1["reject_null"]:
            print(f"   ✅ SIGNIFICANT: p = {test1['p_value']:.2e} < 0.05")
            print(f"   → Shard assignments are NOT random")
        else:
            print(f"   ❌ Not significant: p = {test1['p_value']:.2e}")
    print()
    
    # Test 2: Cross-layer coherence
    print("📊 Test 2: Cross-layer semantic coherence...")
    test2 = compute_cross_layer_coherence()
    results["tests"]["coherence_test"] = test2
    
    if "shard_consistency_score" in test2:
        score = test2["shard_consistency_score"]
        print(f"   Coherence score: {score:.3f}")
        if score > 0.7:
            print(f"   ✅ HIGH coherence - semantic entities stable across layers")
        elif score > 0.5:
            print(f"   ⚠️  MODERATE coherence")
        else:
            print(f"   ❌ LOW coherence")
    print()
    
    # Test 3: Column type clustering
    print("📊 Test 3: Column type semantic clustering...")
    test3 = analyze_column_type_clustering()
    results["tests"]["clustering_test"] = test3
    
    if "p_value" in test3:
        if test3["significant_difference"]:
            print(f"   ✅ SIGNIFICANT: p = {test3['p_value']:.2e} < 0.05")
            print(f"   → Column types cluster differently (semantic meaning)")
        else:
            print(f"   ❌ Not significant: p = {test3['p_value']:.2e}")
    print()
    
    # Test 4: Intra-shard analysis
    print("📊 Test 4: Intra-shard entropy analysis...")
    entropies = []
    for layer in [0, 17, 35, 53, 70]:
        result = compute_intra_shard_similarity(layer)
        if result:
            entropies.append(result["entropy"])
            print(f"   Layer {layer:2d}: entropy = {result['entropy']:.3f}")
    
    if entropies:
        avg_entropy = np.mean(entropies)
        max_entropy = np.log(15)  # Maximum for 15 shards
        normalized = avg_entropy / max_entropy
        
        results["tests"]["entropy_analysis"] = {
            "average_entropy": float(avg_entropy),
            "max_possible_entropy": float(max_entropy),
            "normalized_entropy": float(normalized),
            "interpretation": "Lower entropy = stronger clustering (semantic structure)"
        }
        
        if normalized < 0.8:
            print(f"   ✅ Normalized entropy: {normalized:.3f} < 0.8 (good clustering)")
        else:
            print(f"   ⚠️  Normalized entropy: {normalized:.3f} (weak clustering)")
    print()
    
    # Summary
    print("=" * 70)
    print("📋 SUMMARY")
    print("=" * 70)
    
    significant_tests = 0
    total_tests = 0
    
    for test_name, test_result in results["tests"].items():
        if "p_value" in test_result:
            total_tests += 1
            if test_result.get("reject_null") or test_result.get("significant_difference"):
                significant_tests += 1
    
    print(f"Significant tests: {significant_tests}/{total_tests}")
    print()
    
    if significant_tests >= 2:
        print("✅ CONCLUSION: Statistical evidence supports semantic structure in Monster embeddings")
        print("   The 71-layer, 15-shard vectorization captures real semantic meaning.")
    else:
        print("⚠️  CONCLUSION: Insufficient evidence for semantic structure")
        print("   May need more data or refined analysis.")
    
    print()
    
    # Save results
    with open("semantic_significance_analysis.json", "w") as f:
        json.dump(results, f, indent=2)
    
    print("💾 Results saved to: semantic_significance_analysis.json")
    print()

if __name__ == "__main__":
    main()

File: src/bin/analyze_semantic_significance.rs
// Statistical Significance Analysis of Monster Prime Vectorization
// Proves semantic structure exists in 71-layer embeddings

use polars::prelude::*;
use std::collections::HashMap;
use std::fs::File;
use serde::{Serialize, Deserialize};
use serde_json;

#[derive(Serialize, Deserialize)]
struct ChiSquareTest {
    null_hypothesis: String,
    alternative_hypothesis: String,
    chi_square_statistic: f64,
    p_value: f64,
    reject_null: bool,
    significance_level: f64,
    interpretation: String,
    observed_distribution: Vec<usize>,
    expected_uniform: Vec<f64>,
}

#[derive(Serialize, Deserialize)]
struct CoherenceTest {
    layers_analyzed: Vec<usize>,
    entities_tracked: usize,
    shard_consistency_score: f64,
    interpretation: String,
}

#[derive(Serialize, Deserialize)]
struct EntropyAnalysis {
    average_entropy: f64,
    max_possible_entropy: f64,
    normalized_entropy: f64,
    interpretation: String,
}

#[derive(Serialize, Deserialize)]
struct AnalysisResults {
    analysis_date: String,
    method: String,
    randomness_test: Option<ChiSquareTest>,
    coherence_test: Option<CoherenceTest>,
    entropy_analysis: Option<EntropyAnalysis>,
    conclusion: String,
}

fn load_shard(layer: usize) -> Option<DataFrame> {
    let path = format!("vectors_layer_{:02}.parquet", layer);
    File::open(&path)
        .ok()
        .and_then(|f| ParquetReader::new(f).finish().ok())
}

fn chi_square_test(observed: &[usize], expected: &[f64]) -> (f64, f64) {
    let chi2: f64 = observed.iter().zip(expected.iter())
        .map(|(o, e)| {
            let diff = *o as f64 - e;
            (diff * diff) / e
        })
        .sum();
    
    // Approximate p-value using chi-square distribution (df = 14 for 15 categories)
    let df = 14.0;
    let p_value = 1.0 - chi_square_cdf(chi2, df);
    
    (chi2, p_value)
}

fn chi_square_cdf(x: f64, df: f64) -> f64 {
    // Approximation using gamma function
    if x <= 0.0 { return 0.0; }
    if x > 100.0 { return 1.0; }
    
    // Simple approximation for p-value
    let k = df / 2.0;
    let z = x / 2.0;
    
    // Use series expansion
    let mut sum = 0.0;
    let mut term = 1.0;
    for i in 0..50 {
        sum += term;
        term *= z / (k + i as f64);
        if term < 1e-10 { break; }
    }
    
    let gamma_k = gamma(k);
    1.0 - (z.powf(k) * (-z).exp() * sum / gamma_k).min(1.0)
}

fn gamma(z: f64) -> f64 {
    // Stirling's approximation
    if z < 1.0 { return gamma(z + 1.0) / z; }
    ((2.0 * std::f64::consts::PI / z).sqrt() * (z / std::f64::consts::E).powf(z))
}

fn entropy(counts: &[usize]) -> f64 {
    let total: usize = counts.iter().sum();
    if total == 0 { return 0.0; }
    
    counts.iter()
        .filter(|&&c| c > 0)
        .map(|&c| {
            let p = c as f64 / total as f64;
            -p * p.ln()
        })
        .sum()
}

fn test_randomness() -> Option<ChiSquareTest> {
    println!("📊 Test 1: Testing for non-random structure...");
    
    let mut all_shards = Vec::new();
    
    for layer in 0..71 {
        if let Some(df) = load_shard(layer) {
            if let Ok(col) = df.column("shard") {
                for val in col.u32().unwrap().into_iter() {
                    if let Some(v) = val {
                        all_shards.push(v as usize);
                    }
                }
            }
        }
    }
    
    if all_shards.len() < 100 {
        println!("   ❌ Insufficient data");
        return None;
    }
    
    // Count occurrences
    let mut observed = vec![0usize; 15];
    for &shard in &all_shards {
        if shard < 15 {
            observed[shard] += 1;
        }
    }
    
    // Expected uniform
    let expected: Vec<f64> = vec![all_shards.len() as f64 / 15.0; 15];
    
    let (chi2, p_value) = chi_square_test(&observed, &expected);
    let reject_null = p_value < 0.05;
    
    if reject_null {
        println!("   ✅ SIGNIFICANT: p = {:.2e} < 0.05", p_value);
        println!("   → Shard assignments are NOT random");
    } else {
        println!("   ❌ Not significant: p = {:.2e}", p_value);
    }
    
    Some(ChiSquareTest {
        null_hypothesis: "Shard assignments are uniformly random".to_string(),
        alternative_hypothesis: "Shard assignments follow semantic structure".to_string(),
        chi_square_statistic: chi2,
        p_value,
        reject_null,
        significance_level: 0.05,
        interpretation: "p < 0.05 means semantic structure exists (reject randomness)".to_string(),
        observed_distribution: observed,
        expected_uniform: expected,
    })
}

fn test_coherence() -> Option<CoherenceTest> {
    println!("📊 Test 2: Cross-layer semantic coherence...");
    
    let layers = vec![0, 17, 35, 53, 70];
    let mut dfs = Vec::new();
    
    for &layer in &layers {
        if let Some(df) = load_shard(layer) {
            dfs.push(df);
        }
    }
    
    if dfs.len() < 2 {
        println!("   ❌ Not enough layers");
        return None;
    }
    
    // Track shard consistency per file+column
    let mut entity_shards: HashMap<String, Vec<u32>> = HashMap::new();
    
    for df in &dfs {
        if let (Ok(files), Ok(cols), Ok(shards)) = (
            df.column("file"),
            df.column("column"),
            df.column("shard")
        ) {
            let files = files.str().unwrap();
            let cols = cols.str().unwrap();
            let shards = shards.u32().unwrap();
            
            for i in 0..df.height() {
                if let (Some(f), Some(c), Some(s)) = (files.get(i), cols.get(i), shards.get(i)) {
                    let key = format!("{}::{}", f, c);
                    entity_shards.entry(key).or_insert_with(Vec::new).push(s);
                }
            }
        }
    }
    
    // Compute consistency (low variance = high consistency)
    let mut variances = Vec::new();
    for shards in entity_shards.values() {
        if shards.len() > 1 {
            let mean: f64 = shards.iter().map(|&s| s as f64).sum::<f64>() / shards.len() as f64;
            let var: f64 = shards.iter()
                .map(|&s| (s as f64 - mean).powi(2))
                .sum::<f64>() / shards.len() as f64;
            variances.push(var);
        }
    }
    
    let avg_var = variances.iter().sum::<f64>() / variances.len() as f64;
    let coherence = 1.0 / (1.0 + avg_var);
    
    println!("   Coherence score: {:.3}", coherence);
    if coherence > 0.7 {
        println!("   ✅ HIGH coherence - semantic entities stable across layers");
    } else if coherence > 0.5 {
        println!("   ⚠️  MODERATE coherence");
    } else {
        println!("   ❌ LOW coherence");
    }
    
    Some(CoherenceTest {
        layers_analyzed: layers,
        entities_tracked: entity_shards.len(),
        shard_consistency_score: coherence,
        interpretation: "High score = same semantic entity stays in same Monster prime shard".to_string(),
    })
}

fn test_entropy() -> Option<EntropyAnalysis> {
    println!("📊 Test 3: Intra-shard entropy analysis...");
    
    let test_layers = vec![0, 17, 35, 53, 70];
    let mut entropies = Vec::new();
    
    for &layer in &test_layers {
        if let Some(df) = load_shard(layer) {
            if let Ok(shards) = df.column("shard") {
                let mut counts = vec![0usize; 15];
                for val in shards.u32().unwrap().into_iter() {
                    if let Some(v) = val {
                        if (v as usize) < 15 {
                            counts[v as usize] += 1;
                        }
                    }
                }
                
                let ent = entropy(&counts);
                entropies.push(ent);
                println!("   Layer {:2}: entropy = {:.3}", layer, ent);
            }
        }
    }
    
    if entropies.is_empty() {
        println!("   ❌ No data");
        return None;
    }
    
    let avg_entropy: f64 = entropies.iter().sum::<f64>() / entropies.len() as f64;
    let max_entropy = (15.0f64).ln();
    let normalized = avg_entropy / max_entropy;
    
    if normalized < 0.8 {
        println!("   ✅ Normalized entropy: {:.3} < 0.8 (good clustering)", normalized);
    } else {
        println!("   ⚠️  Normalized entropy: {:.3} (weak clustering)", normalized);
    }
    
    Some(EntropyAnalysis {
        average_entropy: avg_entropy,
        max_possible_entropy: max_entropy,
        normalized_entropy: normalized,
        interpretation: "Lower entropy = stronger clustering (semantic structure)".to_string(),
    })
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🔬 STATISTICAL SIGNIFICANCE ANALYSIS");
    println!("{}", "=".repeat(70));
    println!();
    
    let randomness = test_randomness();
    println!();
    
    let coherence = test_coherence();
    println!();
    
    let entropy_analysis = test_entropy();
    println!();
    
    // Summary
    println!("{}", "=".repeat(70));
    println!("📋 SUMMARY");
    println!("{}", "=".repeat(70));
    
    let mut significant_tests = 0;
    let mut total_tests = 0;
    
    if let Some(ref test) = randomness {
        total_tests += 1;
        if test.reject_null {
            significant_tests += 1;
        }
    }
    
    if coherence.is_some() {
        total_tests += 1;
        if let Some(ref test) = coherence {
            if test.shard_consistency_score > 0.7 {
                significant_tests += 1;
            }
        }
    }
    
    if entropy_analysis.is_some() {
        total_tests += 1;
        if let Some(ref test) = entropy_analysis {
            if test.normalized_entropy < 0.8 {
                significant_tests += 1;
            }
        }
    }
    
    println!("Significant tests: {}/{}", significant_tests, total_tests);
    println!();
    
    let conclusion = if significant_tests >= 2 {
        println!("✅ CONCLUSION: Statistical evidence supports semantic structure in Monster embeddings");
        println!("   The 71-layer, 15-shard vectorization captures real semantic meaning.");
        "Statistical evidence supports semantic structure in Monster embeddings".to_string()
    } else {
        println!("⚠️  CONCLUSION: Insufficient evidence for semantic structure");
        println!("   May need more data or refined analysis.");
        "Insufficient evidence for semantic structure".to_string()
    };
    
    println!();
    
    // Save results
    let results = AnalysisResults {
        analysis_date: "2026-01-29".to_string(),
        method: "Monster Prime Vectorization (71 layers, 15 shards)".to_string(),
        randomness_test: randomness,
        coherence_test: coherence,
        entropy_analysis,
        conclusion,
    };
    
    let json = serde_json::to_string_pretty(&results)?;
    std::fs::write("semantic_significance_analysis.json", json)?;
    
    println!("💾 Results saved to: semantic_significance_analysis.json");
    println!();
    
    Ok(())
}
